include/ccf/pal/mem.h

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include <set>
+#include <stdlib.h>
+
+#if !defined(INSIDE_ENCLAVE) || defined(VIRTUAL_ENCLAVE)
+#  include <cstring>
+#  include <limits>
+#else
+#  include <openenclave/advanced/mallinfo.h>
+#  include <openenclave/bits/security.h>
+#endif
+
+namespace ccf::pal
+{
+  /**
+   * Malloc information formatted based on the OE type, but avoiding to expose
+   * the actual OE type in non-OE code.
+   */
+  struct MallocInfo
+  {
+    size_t max_total_heap_size = 0;
+    size_t current_allocated_heap_size = 0;
+    size_t peak_allocated_heap_size = 0;
+  };
+
+  struct CpuidInfo
+  {
+    uint64_t eax;
+    uint64_t ebx;
+    uint64_t ecx;
+    uint64_t edx;
+  };
+
+  static void cpuid(CpuidInfo* info, uint64_t leaf, uint64_t subleaf)
```

Comment:
I don't think these `cpuid` related types/functions should be in `mem.h`, but instead in `pal/cpuid.h` or `pal/hardware_info.h`.

Reply:
Yeah, I can see that, I think there's also an argument about if getting information about your platform should even be in the abstraction layer since we're not really abstracting it? But that would probably complicate things and I can see that it might not be worth that?

Reply:
I think the next step of the refactor would be to move some of the implementation details to `.cpp` so they're not exposed in the public PAL. Then the user-facing functions like `require_alignment_for_untrusted_reads` (and _maybe_ `is_intel_cpu`) would remain here, but the fact we're actually calling `cpuid` and how we handle that wouldn't be part of the public API.

Reply:
But that's a "maybe one day", not worth doing right now.

Reply:
cpuid is at least partially emulated on at least SGX, and it's not inconceivable it will be on other TEEs too. It's also useful to have a degree of centralised control over this, for targeted manipulation rather than abstraction purposes.

Reply:
Moved to hardware_info.h for now